mtt/src/mtt_types.rs

```diff
@@ -94,27 +94,452 @@ impl MttL2EntryType {
             1 => Some(MttConfidential4K),
             _ => None,
         }
     }
 }
+
+/// MTT related errors.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Error {
+    /// The MTT L2 entry corresponding to the physical address was invalid.
+    /// This should be treated as an operational error.
+    InvalidL2Entry,
+    /// The MTT L1 entry corresponding to the physical address was invalid (encoding of 11b).
+    /// This should be treated as an operational error.
+    InvalidL1Entry,
+    /// The MTT L2 entries for the specified 1GB range failed to pass the invariant that all
+    /// 16 consecutive entries (each mapping 64MB) must have the same type.
+    ///  This should be treated as an operational error.
+    Invalid1GBRange,
+    /// The caller specified an operation on a 1GB range, but the L2 entry for the physical
+    /// address does not map a 1GB range.
+    Non1GBL2Entry,
+    /// The caller specified a operation on a 64MB range, but the L2 entry for the physical
+    /// address doesn't map a 64MB range.
+    Non64MBL2Entry,
+    /// The caller specified a operation on a 4K page, but the L2 entry for the physical
+    /// address doesn't map a 4K page.
+    Non4KL2Entry,
+    /// The caller specified an invalid physical address (too large for example)
+    InvalidAddress,
+    /// The platform initialization hasn't been completed. The implementation is expected
+    /// to initialize pointers to the MTT L2 and L1 page pool using platform-FW tables.
+    PlatformInitNotCompleted,
+    /// The platform didn't allocate sufficient L1 page pool pages to map the specified address
+    InsufficientL1Pages,
+}
+
+/// Holds the result of a MTT operation.
+pub type Result<T> = core::result::Result<T, Error>;
+
+/// Size of the larger memory range containing the page.
+/// This is returned following a successful call to set the type of the page
+/// and can be used for bookkeeping purposes. This can be useful for performance
+/// optimizations like combining or eliding ECALLs to invalidate the MTT.
+pub enum MemoryRangeType {
+    /// The page is part of an 1G non-confidential range
+    NonConfidential1G,
+    /// The page is part of an 1G confidential range
+    Confidential1G,
+    /// The page is part of a 64MB non-confidential range
+    NonConfidential64MB,
+    /// The page is part of a 64MB confidential range
+    Confidential64MB,
+    /// The 4K page is non-confidential
+    NonConfidential4K,
+    /// The 4K page is confidential
+    Confidential4K,
+}
+
+const MTT_L1_ENTRY_COUNT: usize = 512;
+// Struct to represent a L1 page
+// Each entry represents 32 4K pages, which represents a total of 64MB of address space.
+// Each page uses a 2-bit encoding, with '00b as non-confidential and '01b as confidential ('11b is illegal).
+struct MttL1Page {
+    base: [u64; MTT_L1_ENTRY_COUNT],
+}
+struct MttCoreInner {
+    mtt_l2_base: &'static mut [u64],
+    mtt_l1_page_pool_base: &'static mut [MttL1Page],
```

Comment:
As I see it, the platform is responsible for allocating memory for the L1-pages upfront (and marking it as a reserved region ). As such, it's no different from the static allocation for the L2 (albeit, likely with a much larger size on most platforms). However, there may some advantages to have a closure return a page from the pre-allocated pool. For example, the static indexing requires a mapping for the entire address space spanning DRAM, and we may have platforms with large holes in-between. However, wiring this up requires adding code upfront to parse the DT and so forth, and the purpose of this PR was have an initial implementation that allows use to exercise integration with the TEE/R-code.

Reply:
I don't see how you avoid parsing a DT or some other firmware-provided structure for the location of the carveout. Instead what I'm suggesting is to not bake in the assumption into `MttCore` that there's this contiguous chunk of memory that provides sufficient L1 pages for all of physical memory. The closure can simply return successive pages from that carveout for now.

Reply:
I am eager to avoid the on-demand allocation because it would need API and that API would be implementation specific(MTT shouldn't creep in to the API as it's an implementation detail.). But, I agree that bit of policy doesn't need to be enforced at this level and could be handled above.

Having a continuous chunk of memory makes the code a lot simpler, though:)